src/plugins/ui_actions/public/actions/action_factory_definition.ts

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { ActionDefinition } from './action';
+import { Presentable, Configurable } from '../util';
+import { SerializedAction } from './types';
+
+/**
+ * This is a convenience interface for registering new action factories.
+ */
+export interface ActionFactoryDefinition<
+  Config extends object = object,
+  FactoryContext extends object = object,
+  ActionContext extends object = object
+> extends Partial<Presentable<FactoryContext>>, Configurable<Config> {
+  /**
+   * Unique ID of the action factory. This ID is used to identify this action
+   * factory in the registry as well as to construct actions of this ID and
```

Comment:
I've changed to "actions of this type".

I think think we should keep `id`, but `type` is something could also have. The reason why I think we should keep `id` is because this `id` serves 3 things now:

- ID to register in registry. And for all things we are now registering in registry, we consistently are using `id` field, which is nice.
- This ID acts as the `factoryId` for actions.
- We also use this `id` in UI, when we are selecting a factory in React.

In general, I think of `type` more broadly, as I understand, `type` is really a list `types: [ ... ]` as *type* represents the `context` object and some actions can subscribe to multiple triggers, thus receive multiple different `context` objects, e.g.:

```ts
Action<ValueClickContext | RangeBrushContext>
```